rclpy/src/rclpy/handle.cpp

```diff
@@ -0,0 +1,59 @@
+// Copyright 2021 Open Source Robotics Foundation, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include <pybind11/pybind11.h>
+
+#include "rclpy_common/handle.h"
+
+#include "handle.hpp"
+
+namespace rclpy
+{
+Handle::Handle(py::capsule pyhandle)
+: pyhandle_(pyhandle)
+{
+  // Forced to assume this uses rclpy_handle_t because pycapsule name gives no clues
+  auto c_handle = static_cast<rclpy_handle_t *>(pyhandle_);
+
+  // Increment the reference count
+  _rclpy_handle_inc_ref(c_handle);
+}  // namespace rclpy
+
+Handle::~Handle()
+{
+  if (pyhandle_.ptr()) {
+    // Decrement the reference count
+    auto c_handle = static_cast<rclpy_handle_t *>(pyhandle_);
+    _rclpy_handle_dec_ref(c_handle);
+  }
+}
+
+Handle &
+Handle::operator=(const Handle & other)
+{
+  pyhandle_ = other.pyhandle_;
+
```

Comment:
@sloretz hmm, should we decrement the ref count before assignment?

Reply:
`other` will decrement the reference count in its destructor, so I think this function needs to increment the counter to keep it alive.

Reply:
Ohhhh, you mean decrement the ref count of `this->pyhandle_`! Definitely.

Reply:
Decref the old `pyhandle_` before assigning the new one in 1bf8f71

Reply:
Sorry for the imprecise comment, glad you were able to decode me :sweat_smile: